baseplate/core.py

```diff
@@ -238,6 +244,156 @@ class (means that the authentication payload could not be decrypted
 
         return self.payload.get("sub", None)
 
+    @property
+    def user_roles(self):
+        if not self.defined:
+            raise WithheldAuthenticationError
+
+        roles = self.payload.get("user_roles", None)
+        return set(roles) if roles else set()
+
+    @property
+    def oauth_client_id(self):
+        if not self.defined:
+            raise WithheldAuthenticationError
+
+        return self.payload.get("client_id", None)
+
+    @property
+    def oauth_client_type(self):
+        if not self.defined:
+            raise WithheldAuthenticationError
+
+        return self.payload.get("client_type", None)
+
+
+_User = collections.namedtuple(
+    "_User",
+    ["authentication_context", "loid", "cookie_created_ms"],
+)
+OauthClient = collections.namedtuple("OauthClient", ["id", "type"])
+Session = collections.namedtuple("Session", ["id"])
+
+
+class User(_User):
+
+    def event_fields(self):
+        return {
+            "user_id": self.id or self.loid,
+            "user_logged_in": self.is_logged_in,
+            "cookie_created": self.cookie_created_ms,
+        }
+
+    @property
+    def id(self):
+        if self.authentication_context:
+            try:
+                return self.authentication_context.account_id
+            except WithheldAuthenticationError:
+                pass
```

Comment:
Should this and `roles` catch the `WithheldAuthenticationError`s or let those bubble up?  After thinking about it some more, I think I'm leaning towards having them bubble up rather than catching them.

Reply:
Yes, bubble.

Reply:
Same question about the OauthClient stuff?